apache/incubator-netbeans
# [NETBEANS-1807] Fix for PHP Debugging with Chrome ClassNotFoundException Issue

Exception was being thrown when using NB10.0 w/ JDK11 because javax.xml.bind package is no longer included in JDK11.

The java.xml.bind.datatypeconverter class was being used to encode Base64 strings to make authentication keys.  The two files affected were:

* webcommon/netserver/src/org/netbeans/modules/netserver/websocket/WebSocketHandlerClient7.java 
* webcommon/netserver/src/org/netbeans/modules/netserver/websocket/AbstractWSHandler7.java 

This pull request removes the dependency on javax.xml.bind package by using the java.util.Base64 package (available since JDK8) instead.

To Test
--------
Pre-reqs: NB10.0 w/JDK 11 & Google Chrome

1. Install PHP 7.3.1 with XDebug .dll in the /ext directory
2. Create/amend php.ini file with standard XDebug settings
3. Create a new PHP project in NB 10.0 with an index.php file with some PHP (e.g. phpinfo() in the <body>)
4. Amend project properties > run configuration to use PHP Built - in webserver on localhost:9000
5. Start debugging index.php file
6. Step through over statements
6.a Without PR -> as soon as the phpinfo(); statement is stepped over, the Exception will be thrown.  It is silent, so keep an eye on notifications window
6.b With PR -> no exception will be throw.

N.B. Subsequent runs through will not throw Exception - only seems to happen on first pass.

In addition
------------
Whilst investigating this, I discovered that there is another class in NB10.0 that also has a dependency on javax.xml.bind:

* enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/utils/ServerUtils.java 

which appears to be creating basic auth credentials to log into Glassfish.  Haven't amended this file in this PR, as it is a separate issue.
